pkg/pf/tests/diff_set_test.go

```diff
@@ -0,0 +1,243 @@
+					Attributes: map[string]rschema.Attribute{
+						"nested": rschema.StringAttribute{Optional: true},
+					},
+				},
+			},
+		},
+	}
+
+	nestedAttributeReplaceSchema := rschema.Schema{
+		Attributes: map[string]rschema.Attribute{
+			"key": rschema.SetNestedAttribute{
+				Optional: true,
+				NestedObject: rschema.NestedAttributeObject{
+					Attributes: map[string]rschema.Attribute{
+						"nested": rschema.StringAttribute{Optional: true},
+					},
+				},
+				PlanModifiers: []planmodifier.Set{
+					setplanmodifier.RequiresReplace(),
+				},
+			},
+		},
+	}
+
+	nestedAttributeNestedReplaceSchema := rschema.Schema{
+		Attributes: map[string]rschema.Attribute{
+			"key": rschema.SetNestedAttribute{
+				Optional: true,
+				NestedObject: rschema.NestedAttributeObject{
+					Attributes: map[string]rschema.Attribute{
+						"nested": rschema.StringAttribute{
+							Optional: true,
+							PlanModifiers: []planmodifier.String{
+								stringplanmodifier.RequiresReplace(),
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	blockSchema := rschema.Schema{
+		Blocks: map[string]rschema.Block{
+			"key": rschema.SetNestedBlock{
+				NestedObject: rschema.NestedBlockObject{
+					Attributes: map[string]rschema.Attribute{
+						"nested": rschema.StringAttribute{Optional: true},
+					},
+				},
+			},
+		},
+	}
+
+	blockReplaceSchema := rschema.Schema{
+		Blocks: map[string]rschema.Block{
+			"key": rschema.SetNestedBlock{
+				NestedObject: rschema.NestedBlockObject{
+					Attributes: map[string]rschema.Attribute{
+						"nested": rschema.StringAttribute{
+							Optional: true,
+						},
+					},
+				},
+				PlanModifiers: []planmodifier.Set{
+					setplanmodifier.RequiresReplace(),
+				},
+			},
+		},
+	}
+
+	blockNestedReplaceSchema := rschema.Schema{
+		Blocks: map[string]rschema.Block{
+			"key": rschema.SetNestedBlock{
+				NestedObject: rschema.NestedBlockObject{
+					Attributes: map[string]rschema.Attribute{
+						"nested": rschema.StringAttribute{
+							Optional: true,
+							PlanModifiers: []planmodifier.String{
+								stringplanmodifier.RequiresReplace(),
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	attrList := func(arr *[]string) cty.Value {
+		if arr == nil {
+			return cty.NullVal(cty.DynamicPseudoType)
+		}
+		slice := make([]cty.Value, len(*arr))
+		for i, v := range *arr {
+			slice[i] = cty.StringVal(v)
+		}
+		if len(slice) == 0 {
+			return cty.ListValEmpty(cty.String)
+		}
+		return cty.ListVal(slice)
+	}
+
+	nestedAttrList := func(arr *[]string) cty.Value {
+		if arr == nil {
+			return cty.NullVal(cty.DynamicPseudoType)
+		}
+		slice := make([]cty.Value, len(*arr))
+		for i, v := range *arr {
+			slice[i] = cty.ObjectVal(
+				map[string]cty.Value{
+					"nested": cty.StringVal(v),
+				},
+			)
+		}
+		if len(slice) == 0 {
+			return cty.ListValEmpty(cty.Object(map[string]cty.Type{"nested": cty.String}))
+		}
+		return cty.ListVal(slice)
+	}
+
+	schemaValueMakerPairs := []struct {
```

Comment:
What's the algorithm for generating this combination?

If we can't implement it as code, can we at least write it down? I want to know how I can figure out if we are missing a pair.

Reply:
The combination here are of a schema and a valueMaker which can produce a value for that schema.

The schemas are of all the ways TF can express the Set type with RequiresReplaces inserted at each place where it can go.